src/renderers/dom/client/ReactMount.js

```diff
@@ -468,6 +469,14 @@ var ReactMount = {
     var containerHasNonRootReactChild = hasNonRootReactChild(container);
 
     if (__DEV__) {
+      var validElement = typeof nextElement.type === 'string' ?
+      nextElement.type.match(/\-|\s+/) === null &&
+      document.createElement(nextElement.type) instanceof HTMLUnknownElement &&
+      !ReactDOMFactories.hasOwnProperty(nextElement.type) : false;
+
+      warning(!validElement, 'ReactDOM.render(): Invalid component element. ' +
+        'Make sure the component starts with an upper-case letter.');
```

Comment:
I might change the message to something like:

 > 'Unknown element `%s`.  Did you miss-spell an HTML tag name?  If you using a custom component, make sure your custom component starts with an upper-case letter.%s', element.type, ReactComponentTreeDevtool.getStackAddendumByID(debugID)